src/DGtal/base/FunctorHolder.h

```diff
@@ -0,0 +1,291 @@
+/**
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+#pragma once
+
+/**
+ * @file
+ * @author Roland Denis (\c denis@math.univ-lyon1.fr )
+ * CNRS, Institut Camille Jordan (UMR 5208), Université Lyon 1, France
+ *
+ * @date 2018/06/22
+ *
+ * This file is part of the DGtal library.
+ */
+
+#if defined(FunctorHolder_RECURSES)
+#error Recursive header files inclusion detected in FunctorHolder.h
+#else // defined(FunctorHolder_RECURSES)
+/** Prevents recursive inclusion of headers. */
+#define FunctorHolder_RECURSES
+
+#if !defined FunctorHolder_h
+/** Prevents repeated inclusion of headers. */
+#define FunctorHolder_h
+
+//////////////////////////////////////////////////////////////////////////////
+// Inclusions
+#include <iostream>
+#include <type_traits>
+#include <functional>
+#include <memory>
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+namespace functors
+{
+
+//////////////////////////////////////////////////////////////////////////////
+// Implementation details
+namespace
+{
+
+  /// Invoke a callable object without dereferencing it before.
+  template <bool>
+  struct Invoker
+  {
+    template <typename F, typename... T>
+    static inline
+    auto apply(F && fn, T &&... args)
+        -> decltype((std::forward<F>(fn))(std::forward<T>(args)...))
+      {
+        return (std::forward<F>(fn))(std::forward<T>(args)...);
+      }
+  };
+
+  /// Invoke a callable object by dereferencing it before.
```

Comment:
Is this a pointer specialisation ?

Reply:
Exactly, and the choice between those two implementations is actually made when choosing the appropriate storage used in `FunctorHolder` : if we get a lvalue reference, we store the functor by reference and we can call it directly. Otherwise (rvalue reference), we move the functor into a `shared_ptr` and we thus need to dereference it before calling it.

Should I specify it in the code ?